src/Documents/Session/DocumentSession.ts

```diff
@@ -109,43 +116,57 @@ export class DocumentSession implements IDocumentSession {
       return this.requestsExecutor
         .execute(new DeleteDocumentCommand(key, etag))
         .then(() => {
-          PromiseResolver.resolve<IDocument>(document, null, callback);
-          return document;
-        }) as Promise<IDocument>;
+          PromiseResolver.resolve<T>(document, null, callback);
+          return document as T;
+        }) as Promise<T>;
     })
     .catch((error: RavenException) => PromiseResolver.reject(error, null, callback));
   }
 
-  public store(entity: IDocument, documentType?: IDocumentType, key?: DocumentKey, etag?: number, forceConcurrencyCheck?: boolean,
-     callback?: EntityCallback<IDocument>
-  ): Promise<IDocument> {
+  public store<T extends Object>(entity: T, key?: string, etag?: number, forceConcurrencyCheck?: boolean,
+     callback?: EntityCallback<T>
+  ): Promise<T> {
     this.incrementRequestsCount();
 
-    return this.prepareDocumentToStore(entity, documentType, key, etag, forceConcurrencyCheck)
-      .then((entity: IDocument) => {
+    return this.prepareDocumentToStore<T>(entity, key, etag, forceConcurrencyCheck)
+      .then((entity: T) => {
         let tag: number = null;
         const metadata = entity['@metadata'];
-        const conventions: DocumentConventions<IDocument> = this.conventions;
-        const documentKey: DocumentKey = conventions.tryGetIdFromInstance(entity);
+        const conventions: DocumentConventions = this.conventions;
+        const documentKey: string = conventions.tryGetIdFromInstance(entity);
 
         if (conventions.defaultUseOptimisticConcurrency || metadata['force_concurrency_check']) {
           tag = (metadata['@tag'] as number) || conventions.emptyEtag;
         }
 
         return this.requestsExecutor
           .execute(new PutDocumentCommand(documentKey, conventions.tryConvertToRawEntity(entity), tag))
-          .then((): IDocument => {
-            PromiseResolver.resolve<IDocument>(entity, null, callback);
-            return entity;
+          .then((): T => {
+            PromiseResolver.resolve<Object>(entity, null, callback);
+            return entity as T;
           });
       })
       .catch((error: RavenException) => PromiseResolver.reject(error, null, callback));
   }
 
-  public query(documentType?: IDocumentType, indexName?: string, usingDefaultOperator?: boolean, waitForNonStaleResults: boolean = false,
-     includes?: string[], withStatistics: boolean = false
-  ): IDocumentQuery {
-    return new DocumentQuery(this, this.requestsExecutor, documentType);
+  public query<T extends Object>(documentTypeOrObjectType?: string | DocumentConstructor<T>, indexName?: string, options?: IDocumentQueryOptions): IDocumentQuery<T> {
```

Comment:
Since all parameters are optional, perhaps we should wrap everything in opts. I saw in test a load of calls like `session.query(null, null, opts)` and this is not very convenient. Thoughts?